lib/payment/mobilepay_service.dart

```diff
@@ -1,121 +1,103 @@
+import 'package:url_launcher/url_launcher.dart';
 
-  @override
-  void onSuccess(BuildContext context) {
-    //FIXME: use real logic
-    showDialog(
-      context: context,
-      builder: (context) {
-        return const PopupCard(title: 'Success', content: 'Looks good?');
-      },
-    );
-  }
+class MobilePayService implements PaymentHandler {
+  final PurchaseRepository _repository;
 
-  @override
-  void onFailure(BuildContext context) {
-    //FIXME: use real logic
-    showDialog(
-      context: context,
-      builder: (context) {
-        return const PopupCard(title: 'Failure', content: 'Looks good?');
-      },
-    );
-  }
+  MobilePayService(this._repository);
 
   @override
-  void onCancel(BuildContext context) {
-    //FIXME: use real logic
-    showDialog(
-      context: context,
-      builder: (context) {
-        return const PopupCard(title: 'Cancel', content: 'Looks good?');
-      },
-    );
-  }
-
-  @override
-  Future<Payment> initPurchase(int productId) async {
-    // ignore: unused_local_variable
-    final PurchaseRepository _purchaseRepository = sl.get<PurchaseRepository>();
-    // Call coffeecard API with productId
-    //errors:
-    //  networkerror: retry?
-    //  else:         log and report error
-
-    // Receive mobilepayId and deeplink from API
-    //FIXME: call api
-    final InitiatePurchaseResponse response = InitiatePurchaseResponse(
-      id: 122,
-      dateCreated: DateTime.now(),
-      productId: 1,
-      totalAmount: 100,
-      purchaseStatus: 'PendingPayment',
-      paymentDetails: {
-        'paymentType': 'MobilePay',
-        'orderId': 'f5cb3e0f-3b9b-4f50-8c4f-a7450f300a5c',
-        'mobilePayAppRedirectUri':
-            'mobilepay://merchant_payments?payment_id=186d2b31-ff25-4414-9fd1-bfe9807fa8b7',
-        'paymentId': '186d2b31-ff25-4414-9fd1-bfe9807fa8b7'
-      },
-    );
-    //await _purchaseRepositoryinitiatePurchase(productId, PaymentType.mobilepay);
-
-    final Map<String, String> paymentDetails =
-        response.paymentDetails as Map<String, String>;
-
-    return Payment(
-      paymentId: paymentDetails['paymentId']!,
-      deeplink: paymentDetails['mobilePayAppRedirectUri']!,
-    );
+  Future<Either<ApiError, Payment>> initPurchase(int productId) async {
+    final response =
+        await _repository.initiatePurchase(productId, PaymentType.mobilepay);
+
+    if (response is Right) {
+      final purchaseResponse = response.right;
+      final paymentDetails = MobilePayPaymentDetails.fromJsonFactory(
+        purchaseResponse.paymentDetails as Map<String, dynamic>,
+      );
+
+      return Right(
+        Payment(
+          id: purchaseResponse.id!,
+          paymentId: paymentDetails.paymentId!,
+          status: PaymentStatus.awaitingPayment,
+          deeplink: paymentDetails.mobilePayAppRedirectUri!,
+          purchaseTime: purchaseResponse.dateCreated!,
+          price: purchaseResponse.totalAmount!,
+          productName: 'Placeholder',
```

Comment:
do we have a solution to this? Where is this used?

Reply:
The current solution is to not use it. The "fix" resides in the purchase overlay where it is using the name of the product that was passed to it for showing the receipt to the customer, instead of using the one provided by the payment. Personally I would prefer getting the product name from the backend, but it is not really a necessity, as I can get it from the other place

Reply:
is productName required, if not could we omit it then?

Reply:
Currently, it is required, but we could make it nullable if there is a preference. We could also remove it entirely from the object, until we are able to get the information from the backend. I would personally prefer removing it, to making it nullable

Reply:
i think we should remove it for now and introduce it once it is needed

Reply:
Removed